agent/agent_endpoint.go

```diff
@@ -190,31 +190,203 @@ func (s *HTTPServer) AgentServices(resp http.ResponseWriter, req *http.Request)
 			EnableTagOverride: s.EnableTagOverride,
 			CreateIndex:       s.CreateIndex,
 			ModifyIndex:       s.ModifyIndex,
-			ProxyDestination:  s.ProxyDestination,
 			Weights:           weights,
 		}
+
 		if as.Tags == nil {
 			as.Tags = []string{}
 		}
 		if as.Meta == nil {
 			as.Meta = map[string]string{}
 		}
-		// Attach Connect configs if the exist
+		// Attach Unmanaged Proxy config if exists
+		if s.Kind == structs.ServiceKindConnectProxy {
+			as.Proxy = s.Proxy.ToAPI()
+			// DEPRECATED (ProxyDestination) - remove this when removing ProxyDestination
+			// Also set the deprecated ProxyDestination
+			as.ProxyDestination = as.Proxy.DestinationServiceName
+		}
+
+		// Attach Connect configs if the exist. We use the actual proxy state since
+		// that may have had defaults filled in compared to the config that was
+		// provided with the service as stored in the NodeService here.
 		if proxy, ok := proxies[id+"-proxy"]; ok {
 			as.Connect = &api.AgentServiceConnect{
 				Proxy: &api.AgentServiceConnectProxy{
-					ExecMode: api.ProxyExecMode(proxy.Proxy.ExecMode.String()),
-					Command:  proxy.Proxy.Command,
-					Config:   proxy.Proxy.Config,
+					ExecMode:  api.ProxyExecMode(proxy.Proxy.ExecMode.String()),
+					Command:   proxy.Proxy.Command,
+					Config:    proxy.Proxy.Config,
+					Upstreams: proxy.Proxy.Upstreams.ToAPI(),
 				},
 			}
+		} else if s.Connect.Native {
+			as.Connect = &api.AgentServiceConnect{
+				Native: true,
+			}
 		}
 		agentSvcs[id] = as
 	}
 
 	return agentSvcs, nil
 }
 
+// GET /v1/agent/service/:service_id
+//
+// Returns the service definition for a single local services and allows
+// blocking watch using hash-based blocking.
+func (s *HTTPServer) AgentService(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	// Get the proxy ID. Note that this is the ID of a proxy's service instance.
+	id := strings.TrimPrefix(req.URL.Path, "/v1/agent/service/")
```

Comment:
Service IDs and names are a mess - they have zero validation historically and now cause horrible problems just about everywhere. We haven't bitten the bullet to add sane validation to them because it will break someone somewhere but note that we have many many endpoints that assume service ids and names are url safe already so I don't think this makes the mess any worse :(

It's been a TODO item for the longest time but TODO items that involve coordinating a significant breaking change seem to always be better done later...